sound/soc/sof/ipc4-topology.h

```diff
@@ -359,4 +359,39 @@ struct sof_ipc4_process {
 	bool payload_with_output_fmt;
 };
 
+/**
+ * struct sof_ipc4_pin_format - widgets pin format
+ * @pin_index: input/output pin index for the widget
+ * @iobs: input/output buffer size for input/output pin
+ * @audio_fmt: audio format for the pin
+ *
+ * This structure can be used for both source or sink pin cases,
+ * and the pin_index is relative to the type of pin being used.
+ */
+struct sof_ipc4_pin_format {
+	u32 pin_index;
+	u32 iobs;
+	struct sof_ipc4_audio_format audio_fmt;
+};
+
+/**
+ * struct sof_ipc4_base_module_cfg_ext - base module config extension for loadable module
+ * @num_sink_pin_fmts: number of sink pin formats in the data member
+ * @num_source_pin_fmts: number of source pin formats in the data member
+ * @reserved: reserved for future use
+ * @priv_param_len: length of optional module specific parameters in the data member
+ * @data: consist of three parts:
+ *        a) num_sink_pin_fmts * sof_ipc4_pin_format followed by
+ *        b) num_source_pin_fmts * sof_ipc4_pin_format, and then followed by
+ *        c) priv_param_len bytes of module specific parameters.
+ */
+struct sof_ipc4_base_module_cfg_ext {
+	/* Loadable modules have at least one sink pin and one source pin */
+	u16 num_sink_pin_fmts;
```

Comment:
The built-in modules and external modules differ in the initialization payload format:
- build-in modules: base_module_config + optional output format(exist if payload_with_output_fmt token is true) + optional module specific extension(exist if process->cdata != NULL)
- external modules: base_module_config + base_module_config_extension

struct sof_ipc4_base_module_cfg_ext is designed for external modules in the avs firmware, No built-in module uses it in the avs firmware. I think we should align the behavior with avs firmware and its external modules.

If we want to migrate a built-in module to external library, we should migrate the payload format(which should be easy), or it will not work both in Linux and windows OSs.

Reply:
@aiChaoSONG @ujfalusi Looking at the documents and I think we don't need to discuss these in terms of built-in versus external. The "base_module_config + base_module_config_extension" is a standard interface available for all newer modules.